src/resources/common.ts

```diff
@@ -1,41 +1,105 @@
+// Copyright 2021, Nitric Technologies Pty Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import {
+  Resource as ProtoResource,
+  ResourceType,
+  PolicyResource,
+  ResourceDeclareRequest,
+  ResourceDeclareResponse,
+  ActionMap,
+} from '@nitric/api/proto/resource/v1/resource_pb';
+import resourceClient from './client';
+
+export type ActionsList = ActionMap[keyof ActionMap][];
+
 export abstract class Resource {
   /**
    * Unique name for the resource by type within the stack.
    *
    * This name can be used in all future references, it will be resolved automatically at runtime.
    */
   protected readonly name: string;
+  /**
+   * Used to resolve the given reource for policy creation
+   */
+  public registerPromise: Promise<ProtoResource>;
 
   constructor(name: string) {
     this.name = name;
   }
 
-  protected abstract register(): Promise<void>
+  public registerPolicy(...perms: string[]) {
+    const actions = this.permsToActions(...perms);
+
+    const req = new ResourceDeclareRequest();
+    const policyResource = new ProtoResource();
+    policyResource.setType(ResourceType.POLICY);
+
+    const policy = new PolicyResource();
+    policy.setActionsList(actions);
+
+    req.setResource(policyResource);
+    req.setPolicy(policy);
+
+    this.registerPromise.then((resource) => {
+      policy.setResourcesList([resource]);
+
+      resourceClient.declare(
+        req,
+        (error, response: ResourceDeclareResponse) => {
+          if (error) {
+            // TODO: remove this ignore when not using link
+            // @ts-ignore
+            throw new Error(fromGrpcError(error));
+          }
+        }
+      );
+    });
+  }
+
+  protected abstract register(): Promise<ProtoResource>;
+
+  protected abstract permsToActions(...perms: string[]): ActionsList;
```

Comment:
One more thing, I would still make this generic, then we get type saftey in the underlying implementations which means we get autocomplete and it will make typos a compile error.